frmts/vrt/vrtexpression.cpp

```diff
@@ -0,0 +1,224 @@
+#include "cpl_conv.h"
+#include "cpl_error.h"
+#include "cpl_string.h"
+#include "vrtexpression.h"
+
+#define exprtk_disable_caseinsensitivity
+#define exprtk_disable_rtl_io
+#define exprtk_disable_rtl_io_file
+#define exprtk_disable_rtl_vecops
+#define exprtk_disable_string_capabilities
+#include <exprtk.hpp>
+
+#include <cstdint>
+#include <sstream>
+
+struct vector_access_check final : public exprtk::vector_access_runtime_check
+{
+    bool handle_runtime_violation(violation_context &context) override
+    {
+        auto nElements = (static_cast<std::uint8_t *>(context.end_ptr) -
+                          static_cast<std::uint8_t *>(context.base_ptr)) /
+                         context.type_size;
+        auto nIndexAccessed = (static_cast<std::uint8_t *>(context.access_ptr) -
+                               static_cast<std::uint8_t *>(context.base_ptr)) /
+                              context.type_size;
+
+        std::ostringstream oss;
+        oss << "Attempted to access index " << nIndexAccessed
+            << " in a vector of " << nElements
+            << " elements when evaluating VRT expression.";
+        throw std::runtime_error(oss.str());
+    }
+};
+
+class GDALExpressionEvaluator::Impl
+{
+  public:
+    exprtk::expression<double> m_oExpression{};
+    exprtk::parser<double> m_oParser{};
+    exprtk::symbol_table<double> m_oSymbolTable{};
+    std::string m_osExpression{};
+
+    std::vector<std::pair<std::string, double *>> m_aoVariables{};
+    std::vector<std::pair<std::string, std::vector<double> *>> m_aoVectors{};
+    std::vector<double> m_adfResults{};
+    vector_access_check m_oVectorAccessCheck{};
+
+    bool m_bIsCompiled{false};
+
+    Impl()
+    {
+        using settings_t = std::decay_t<decltype(m_oParser.settings())>;
+
+        m_oParser.register_vector_access_runtime_check(m_oVectorAccessCheck);
+
+        int nMaxVectorSize = std::atoi(
+            CPLGetConfigOption("GDAL_EXPRTK_MAX_VECTOR_SIZE", "100000"));
```

Comment:
maybe use your new CPLParseMemorySize ?

Reply:
We're limiting the number of elements rather than the memory size. I've renamed the config option to make that less confusing.